File: 2958-length-of-longest-subarray-with-at-most-k-frequency/2958-length-of-longest-subarray-with-at-most-k-frequency.cpp
class Solution {
public:
    int maxSubarrayLength(vector<int>& nums, int k) {
        int n = nums.size();
        unordered_map<int, int> mp;
        int i=0, j=0;
        int maxLen = 0;
        while(j<n){
            mp[nums[j]]++;
            if(mp[nums[j]] <= k){
                maxLen = max(maxLen, j-i+1);
            }
            while(mp[nums[j]] > k){
                mp[nums[i]]--;
                i++;
            }
            j++;
        }
        return maxLen;
    }
};

File: 2958-length-of-longest-subarray-with-at-most-k-frequency/2958-length-of-longest-subarray-with-at-most-k-frequency.java
class Solution {
    public int maxSubarrayLength(int[] nums, int k) {
        int n = nums.length;
        HashMap<Integer, Integer> mp = new HashMap<>();
        int i=0, j=0;
        int maxLen = 0;
        
        while(j<n){
            mp.put(nums[j], mp.getOrDefault(nums[j], 0)+1);
            while(mp.get(nums[j]) > k){
                mp.put(nums[i], mp.get(nums[i])-1);
                i++;
            }
            maxLen = Math.max(maxLen, j-i+1);
            j++;
        }
        return maxLen;
    }
}